beacon-chain/sync/validate_aggregate_proof.go

```diff
@@ -210,32 +224,74 @@ func validateIndexInCommittee(ctx context.Context, bs *stateTrie.BeaconState, a
 	return nil
 }
 
-// This validates selection proof by validating it's from the correct validator index of the slot and selection
-// proof is a valid signature.
-func validateSelection(ctx context.Context, bs *stateTrie.BeaconState, data *ethpb.AttestationData, validatorIndex uint64, proof []byte) error {
-	_, span := trace.StartSpan(ctx, "sync.validateSelection")
+// This validates selection proof by validating it's from the correct validator index of the slot.
+// It does not verify the selection proof, it returns the signature set of selection proof which can be used for batch verify.
+func validateSelectionIndex(ctx context.Context, bs *stateTrie.BeaconState, data *ethpb.AttestationData, validatorIndex uint64, proof []byte) (*bls.SignatureSet, error) {
+	_, span := trace.StartSpan(ctx, "sync.validateSelectionIndex")
 	defer span.End()
 
 	committee, err := helpers.BeaconCommitteeFromState(bs, data.Slot, data.CommitteeIndex)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	aggregator, err := helpers.IsAggregator(uint64(len(committee)), proof)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if !aggregator {
-		return fmt.Errorf("validator is not an aggregator for slot %d", data.Slot)
+		return nil, fmt.Errorf("validator is not an aggregator for slot %d", data.Slot)
 	}
 
 	domain := params.BeaconConfig().DomainSelectionProof
 	epoch := helpers.SlotToEpoch(data.Slot)
-	return helpers.ComputeDomainVerifySigningRoot(bs, validatorIndex, epoch, data.Slot, domain, proof)
+
+	v, err := bs.ValidatorAtIndex(validatorIndex)
+	if err != nil {
+		return nil, err
+	}
+	publicKey, err := bls.PublicKeyFromBytes(v.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	d, err := helpers.Domain(bs.Fork(), epoch, domain, bs.GenesisValidatorRoot())
+	if err != nil {
+		return nil, err
+	}
+	root, err := helpers.ComputeSigningRoot(data.Slot, d)
+	if err != nil {
+		return nil, err
+	}
+	return &bls.SignatureSet{
+		Signatures: [][]byte{proof},
+		PublicKeys: []bls.PublicKey{publicKey},
+		Messages:   [][32]byte{root},
+	}, nil
 }
 
-// This verifies aggregator signature over the signed aggregate and proof object.
-func validateAggregatorSignature(s *stateTrie.BeaconState, a *ethpb.SignedAggregateAttestationAndProof) error {
-	return helpers.ComputeDomainVerifySigningRoot(s, a.Message.AggregatorIndex,
-		helpers.SlotToEpoch(a.Message.Aggregate.Data.Slot), a.Message, params.BeaconConfig().DomainAggregateAndProof, a.Signature)
+// This returns aggregator signature set which can be used to batch verify.
+func aggSigSet(s *stateTrie.BeaconState, a *ethpb.SignedAggregateAttestationAndProof) (*bls.SignatureSet, error) {
```

Comment:
can we use our core methods here if possible, the logic here seems duplicated.

Reply:
which core method? this is for `DomainAggregateAndProof` over the entire `message`

Reply:
wouldn't we just be using this ?
```go
func validateAggregatorSignature(s *stateTrie.BeaconState, a *ethpb.SignedAggregateAttestationAndProof) error {
	return helpers.ComputeDomainVerifySigningRoot(s, a.Message.AggregatorIndex,
		helpers.SlotToEpoch(a.Message.Aggregate.Data.Slot), a.Message, params.BeaconConfig().DomainAggregateAndProof, a.Signature)
}
```
any way to refactor that method to return a signature set instead ? Goal would be to avoid having 2 different methods
for signature verification.

Reply:
I'll open a subsequent PR to address this. We should just refactor for all. There's some duplications from BLS helpers to signing helpers to block processing and to this. Better to improve all at once